### ISSUE TITLE
Beckhoff EP2316 does not work

### ISSUE THREAD

**Issue by Bouni** [linuxcnc-ethercat/linuxcnc-ethercat#11]
I had a Beckhoff [EP2316-0003](https://www.beckhoff.com/en-en/products/i-o/ethercat-box/epxxxx-industrial-housing/ep2xxx-digital-output/ep2316-0003.html) at hand and thought I'll try if it works because there's a [comment in the code](https://github.com/linuxcnc-ethercat/linuxcnc-ethercat/blob/master/src/lcec_ep23xx.h#L28) saying "Untested, may not work".

Unfortunately it does not :pensive:

```
Debug file information:
Note: Using POSIX realtime
Failed to register PDO entry: No such file or directory
LCEC: master 0 PDO entry registration failed
lcec: rtapi_app_main: Invalid argument (-22)
./ethercat.hal:2: waitpid failed /usr/bin/rtapi_app lcec
./ethercat.hal:2: /usr/bin/rtapi_app exited without becoming ready
./ethercat.hal:2: insmod for lcec failed, returned -1
1779
Stopping realtime threads
Unloading hal components
Note: Using POSIX realtime
```

```
<masters>
<master idx="0" appTimePeriod="1000000" refClockSyncCycles="5">
      <slave idx="0" type="EL3255" />
      <slave idx="1" type="EL3255" />
      <slave idx="2" type="EL1018" />
      <slave idx="3" type="EL1018" />
      <slave idx="4" type="EL2008" />
      <slave idx="5" type="EK1122" />
      <slave idx="6" type="EP2316" />
</master>
</masters>
```

Any ideas why this does not work?

**Comment by Bouni** [comment]
I managed to get the EP2316 working as a generic device using this config:

```
<masters>
<master idx="0" appTimePeriod="1000000" refClockSyncCycles="5">
  <slave idx="0" type="EK1122" />
  <slave idx="1" type="generic" vid="00000002" pid="090C4052" configPdos="false">
    <syncManager idx="0" dir="out">
      </syncManager>
      <syncManager idx="1" dir="in">
      </syncManager>
      <syncManager idx="2" dir="out">
        <pdo idx="1600">
          <pdoEntry idx="7000" subIdx="01" bitLen="1" halPin="out-0" halType="bit"/>
          <pdoEntry idx="7000" subIdx="02" bitLen="1" halPin="out-1" halType="bit"/>
          <pdoEntry idx="7000" subIdx="03" bitLen="1" halPin="out-2" halType="bit"/>
          <pdoEntry idx="7000" subIdx="04" bitLen="1" halPin="out-3" halType="bit"/>
          <pdoEntry idx="7000" subIdx="05" bitLen="1" halPin="out-4" halType="bit"/>
          <pdoEntry idx="7000" subIdx="06" bitLen="1" halPin="out-5" halType="bit"/>
          <pdoEntry idx="7000" subIdx="07" bitLen="1" halPin="out-6" halType="bit"/>
          <pdoEntry idx="7000" subIdx="08" bitLen="1" halPin="out-7" halType="bit"/>
        </pdo>
      </syncManager>
      <syncManager idx="3" dir="in">
        <pdo idx="1a00">
          <pdoEntry idx="6000" subIdx="01" bitLen="1" halPin="in-0" halType="bit"/>
          <pdoEntry idx="6000" subIdx="02" bitLen="1" halPin="in-1" halType="bit"/>
          <pdoEntry idx="6000" subIdx="03" bitLen="1" halPin="in-2" halType="bit"/>
          <pdoEntry idx="6000" subIdx="04" bitLen="1" halPin="in-3" halType="bit"/>
          <pdoEntry idx="6000" subIdx="05" bitLen="1" halPin="in-4" halType="bit"/>
          <pdoEntry idx="6000" subIdx="06" bitLen="1" halPin="in-5" halType="bit"/>
          <pdoEntry idx="6000" subIdx="07" bitLen="1" halPin="in-6" halType="bit"/>
          <pdoEntry idx="6000" subIdx="08" bitLen="1" halPin="in-7" halType="bit"/>
        </pdo>
      </syncManager>
    </slave>
</master>
</masters>
```

Maybe this helps figuring out why its not working with `type="EP2316"`

**Comment by Bouni** [comment]
I figured it out 🥳 

The `EP2338` and the `EP2349` have PDOs like 6000, 6010, 6020 with a fix subindex and so on where the EP2316 has just 6000 and an incrementing subindex.

I'll send a PR within the next minutes.

**Comment by Bouni** [comment]
Closed #12 in favor of #15
